src/updates.c

```diff
@@ -0,0 +1,706 @@
+
+  if (!flatpak_transaction_run (transaction, NULL, &error))
+    {
+      g_warning ("Transaction failed: %s", error->message);
+      goto out;
+    }
+
+  response = 0;
+
+out:
+  if (request->exported)
+    {
+      xdp_request_emit_response (XDP_REQUEST (request), response, g_variant_builder_end (&opt_builder));
+      request_unexport (request);
+    }
+
+  updates_data_free (d);
+}
+
+static gboolean
+handle_install_update (XdpUpdates *object,
+                       GDBusMethodInvocation *invocation,
+                       const char *arg_window,
+                       GVariant *arg_options)
+{
+  Request *request = request_from_invocation (invocation);
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GTask) task = NULL;
+
+  REQUEST_AUTOLOCK (request);
+
+  g_object_set_data_full (G_OBJECT (request), "window", g_strdup (arg_window), g_free);
+
+  request_export (request, g_dbus_method_invocation_get_connection (invocation));
+
+  task = g_task_new (object, NULL, NULL, NULL);
+  g_task_set_task_data (task, g_object_ref (request), g_object_unref);
+  g_task_run_in_thread (task, handle_install_update_in_thread_func);
+
+  xdp_updates_complete_install_update (object, invocation, request->id);
+
+  return TRUE;
+}
+
+typedef struct _UpdatesSession
+{
+  Session parent;
+  char *ref;
+  FlatpakInstallation *installation;
+  guint timeout;
+
+  gboolean closed;
+} UpdatesSession;
+
+typedef struct _UpdatesSessionClass
+{
+  SessionClass parent_class;
+} UpdatesSessionClass;
+
+GType updates_session_get_type (void);
+
+G_DEFINE_TYPE (UpdatesSession, updates_session, session_get_type ())
+
+static void
+updates_session_close (Session *session)
+{
+  UpdatesSession *updates_session = (UpdatesSession *)session;
+
+  updates_session->closed = TRUE;
+
+  g_debug ("updates session owned by '%s' closed", session->sender);
+}
+
+static void
+updates_session_finalize (GObject *object)
+{
+  UpdatesSession *session = (UpdatesSession*)object;
+
+  if (session->timeout)
+    g_source_remove (session->timeout);
+  g_free (session->ref);
+  g_clear_object (&session->installation);
+
+  G_OBJECT_CLASS (updates_session_parent_class)->finalize (object);
+}
+
+static void
+updates_session_init (UpdatesSession *session)
+{
+}
+
+static void
+updates_session_class_init (UpdatesSessionClass *klass)
+{
+  GObjectClass *object_class;
+  SessionClass *session_class;
+
+  object_class = G_OBJECT_CLASS (klass);
+  object_class->finalize = updates_session_finalize;
+
+  session_class = (SessionClass *)klass;
+  session_class->close = updates_session_close;
+}
+
+static gboolean
+check_for_updates (gpointer data)
+{
+  Session *session = data;
+  UpdatesSession *usession = data;
+  g_autoptr(GPtrArray) updates = NULL;
+  int i;
+
+  g_debug ("Checking for updates for %s", usession->ref);
+
+  updates = flatpak_installation_list_installed_refs_for_update (usession->installation, NULL, NULL);
```

Comment:
This is a bit overkill as it will scan all remotes and stuff. I think what you want is:

installed_ref = flatpak_installation_get_installed_ref(installation, ...);
remote_ref = (flatpak_installation_fetch_remote_ref_sync(installation, flatpak_installed_ref_get_origin (installed_ref), ...);
should_update = g_strcmp0 (flatpak_ref_get_commit (installed_ref, remote_ref));

In fact, I'm not quite sure this is right either. flatpak_installation_get_installed_ref() will get you the latest *installed* version. But maybe you want to look at the app-commit= line in the .flatpak-info (i.e. the actual running version)? Its a bit weird, as it doesn't fit the notion of "update", but maybe the API should do:
 * If there is a newer version remote than what is installed, install that
 * If there is a newer version installed than what is running, pretend to install that and immediately succeed.
 * Otherwise do nothing.

Its a bit unclear really what is the expected behaviour here if we start racing with e.g. gnome-software doing an update in the background.

Reply:
Maybe we should add a flag to the signal actually, whether it needs installing or if its already installed.